compiler/rustc_lint/src/builtin.rs

```diff
@@ -992,36 +992,69 @@ declare_lint! {
     "generic items must be mangled"
 }
 
-declare_lint_pass!(InvalidNoMangleItems => [NO_MANGLE_CONST_ITEMS, NO_MANGLE_GENERIC_ITEMS]);
+declare_lint! {
+    /// The `export_name_generic_items` lint detects using `#[export_name]` on generic functions.
+    ///
+    /// ### Example
+    /// ```rust,compile_fail
+    /// #[unsafe(export_name = "exported_name")]
+    /// fn foo<T>() {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// A function with generics must have its symbol mangled to accommodate
+    /// the generic parameter. The [`export_name` attribute] has no effect in
+    /// this situation, and should be removed.
+    ///
+    /// [`export_name` attribute]: https://doc.rust-lang.org/reference/abi.html#the-export_name-attribute
+    pub EXPORT_NAME_GENERIC_ITEMS,
+    Warn,
+    "generic items must be mangled, export_name would avoid mangling"
+}
```

Comment:
Does this actually need to be a separate lint? `#[unsafe(no_mangle)] fn foo()` is pretty much an alias for `#[unsafe(export_name = "foo")] fn foo()`.